### ISSUE TITLE
OutdatedDocumentation is too vague

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5561]
## Context
Whenever I see `OutdatedDocumentation` violation, I'm always baffled as to what to do. Over the months of using it I kind-of learned that it has to do with the method parameters, vs the KDoc, but it's still a lot of thinking to figure out what needs changing.

## Expected Behavior of the rule
To fix this, I propose to change `OutdatedDocumentation.reportIfDocumentationIsOutdated` to not only query the declarations with a `Boolean`, but actually grab a pair of `(extra: List<Declaration>, missing: List<Declaration>)` from `declarationsMatch` from the perspective of KDoc (because of the rule name). From this the outer function can detect if there's a mismatch and provide a nice error message.

## Example
Code
```kotlin
/**
 * ...
 * @param <T> the static type accepted by the matcher
 * (This can flag obvious compile-time problems such as `assumeThat(1, is("a"))`.)
 * @param actual the computed value being compared
 * @param matcher an expression, built of [Matcher]s, specifying allowed values
 *
 * @see org.hamcrest.Matcher
 * @see org.hamcrest.Matchers
 */
fun <T> assumeThat(message: String?, actual: T, matcher: Matcher<in T>) {
```
Original:
```
org\hamcrest\MatcherAssume.kt:72:9: Documentation of assumeThat is outdated [OutdatedDocumentation]
```
Expected:
```
org\hamcrest\MatcherAssume.kt:72:9: Documentation of assumeThat is outdated, documentation for message parameter is missing. [OutdatedDocumentation]
```

Combining multiple problems into one message is left to the implementor.

---

Alternatively the reporting can be done on a per-parameter level, instead of method, but that might be a bigger change and doesn't cover when there's extra documentation entry for a non-existent parameter.

**Comment by cortinico** [comment]
> Alternatively the reporting can be done on a per-parameter level,

Yes I believe that's the way forward as it's the closest to anything actionable for the user.

**Comment by TWiStErRob** [comment]
> but that [...] doesn't cover when there's extra documentation entry for a non-existent parameter.
```kotlin
/**
 * @param extra foo
 * @param existing bar
 */
fun f(existing: String)
```

**Comment by BraisGabin** [comment]
I agree, the error messages should be as actionable as possible. And I think that for your last example the message should be similar.

**Comment by atulgpt** [comment]
Hi @BraisGabin, @TWiStErRob how should we handle the following cases?

1. When there are multiple violations? Should we just append all other errors? Like `Documentation of assumeThat is outdated, documentation for message parameter is missing, documentation of throwable is not needed, documentation of stack trace is missing.`
2. For cases when `matchDeclarationsOrder` is true and documentation of each property is present but is out of order. Should we follow the error reporting like  \# 1 where we report each property as no property will match anything? Is there any recommendation to shorten the error msg in this case?
3. Should we update the error message in case `matchDeclarationsOrder` is true to something like `Documentation of assumeThat is outdated or misordered, <error msg per mismatch>...`?

**Comment by BraisGabin** [comment]
I think that in general detekt reports one issue for each violation. So I would raise one issue for each missing parameter. And about the ordering, we could use the same as the rule that defines the order of the classes (I can't recall the name now) to minimize the number of errors.

The problem is the name of the rule "outdated documentation". It's strange to have multiple outdated errors on the same method. But I would vote to mark this rule to rename for 2.0.

**Comment by atulgpt** [comment]
Hi, @BraisGabin so apart from renaming, reporting multiple issues(for each missing property or additional property and if `matchDeclarationsOrder` is set then each reorder required) can be picked right now. Right? I am willing to fix this issue if my understanding is correct.

**Comment by BraisGabin** [comment]
Yes, I think so.

**Comment by TWiStErRob** [comment]
Things said above make sense to me.

If the violations target individual parameters (which have missing docs), the existing suppression will still work because of scoping, but the baselines will need regenerating, but I guess this is expected for every release, right?

For extra kdoc `@param` is it possible to target the `@param paramname` part of the KDoc with a violation? That would be extra awesome, because it's very much in context. I'm not sure if suppression works the same way as for function parameters.

For ordering, I think it's enough to just say `Parameter documentation order is mismatched with declaration.` and put the violation on the function name. There's no need to say which one is out of order, as it should be quite clear, unless the parameter names are very similar like described in [gradle wrapper validation action docs](https://github.com/gradle/wrapper-validation-action#solution).

These 3 together would mean that there cannot be two problems on one code element, and it would be very clear where the problem is. Assuming of course part of KDoc can be flagged.

**Comment by BraisGabin** [comment]
Yes, we don't care about breaking the baseline. I mean, if possible is better to don't break it but if there is a improvement in other point it is worth it. I mean, each time we resolve a false negative we mess with it too.
